src/SqlSquared/Parser.purs

```diff
@@ -60,314 +88,227 @@ parseModule = tokenize >=> flip P.runParser (go <* eof)
 
-operator ∷ ∀ m. Monad m ⇒ String →  P.ParserT (Array Token) m Unit
-operator s = PC.try $ token >>= case _ of
+operator ∷ ∀ m. Monad m ⇒ String →  P.ParserT TokenStream m Unit
+operator s = withToken (printToken (Op s)) case _ of
   Op ss | s == ss → pure unit
-  _ → P.fail $  "Token is not an operator " <> s
+  tok → P.fail $ printToken tok
 
-keyword ∷ ∀ m. Monad m ⇒ String →  P.ParserT (Array Token) m String
-keyword s = PC.try $ token >>= case _ of
+keyword ∷ ∀ m. Monad m ⇒ String →  P.ParserT TokenStream m String
+keyword s = withToken (printToken (Kw s)) case _ of
   Kw ss | S.toLower s == S.toLower ss → pure s
-  _ → P.fail $ "Token is not a keyword " <> s
+  tok → P.fail $ printToken tok
 
-anyKeyword ∷ ∀ m. Monad m ⇒ P.ParserT (Array Token) m String
-anyKeyword = PC.try $ token >>= case _ of
+anyKeyword ∷ ∀ m. Monad m ⇒ P.ParserT TokenStream m String
+anyKeyword = withToken "keyword" case _ of
   Kw s → pure s
-  _ → P.fail "Token is not a keyword"
-
-match ∷ ∀ m. Monad m ⇒ Token → P.ParserT (Array Token) m Token
-match = whenTok ∘ eq
-
-expr
-  ∷ ∀ t m
-  . Corecursive t (Sig.SqlF EJ.EJsonF)
-  ⇒ Monad m
-  ⇒ P.ParserT (Array Token) m t
-expr = letExpr
-
-letExpr
-  ∷ ∀ t m
-  . Corecursive t (Sig.SqlF EJ.EJsonF)
-  ⇒ Monad m
-  ⇒ P.ParserT (Array Token) m t
-letExpr = let_ <|> queryExpr
-  where
-  let_ = PC.try do
-    i ← ident
-    operator ":="
-    bindTo ← expr
-    operator ";"
-    in_ ← expr
-    pure $ embed $ Sig.Let { ident: i, bindTo, in_ }
-
-queryExpr
-  ∷ ∀ t m
-  . Corecursive t (Sig.SqlF EJ.EJsonF)
-  ⇒ Monad m
-  ⇒ P.ParserT (Array Token) m t
-queryExpr = PC.try do
-  prod (query <|> definedExpr) queryBinop \lhs rhs op →
-    embed $ Sig.Binop { lhs, rhs, op }
-
-
-queryBinop ∷ ∀ m. Monad m ⇒ P.ParserT (Array Token) m Sig.BinaryOperator
-queryBinop =
-  (keyword "limit" $> Sig.Limit)
-  <|> (keyword "offset" $> Sig.Offset)
-  <|> (keyword "sample" $> Sig.Sample)
-  <|> (keyword "union" $> Sig.Union)
-  <|> (keyword "union" *> keyword "all" $> Sig.UnionAll)
-  <|> (keyword "intersect" $> Sig.Intersect)
-  <|> (keyword "intersect" *> keyword "all" $> Sig.IntersectAll)
-  <|> (keyword "except" $> Sig.Except)
-
+  tok → P.fail $ printToken tok
+
+expr ∷ ∀ m t. SqlParser' m t
+expr = letExpr <|> queryExpr
+
+letExpr ∷ ∀ m t. SqlParser' m t
+letExpr = do
+  i ← PC.try (ident <* operator ":=")
+  bindTo ← expr
+  operator ";"
+  in_ ← expr
+  pure $ embed $ Sig.Let { ident: i, bindTo, in_ }
+
+queryExpr ∷ ∀ m t. SqlParser' m t
+queryExpr = prod (query <|> definedExpr) queryBinop _BINOP
+
+queryBinop ∷ ∀ m. Monad m ⇒ P.ParserT TokenStream m Sig.BinaryOperator
+queryBinop = asErrorMessage "query operator" $ PC.choice
+  [ keyword "limit" $> Sig.Limit
+  , keyword "offset" $> Sig.Offset
+  , keyword "sample" $> Sig.Sample
+  , keyword "union" $> Sig.Union
+  , keyword "union" *> keyword "all" $> Sig.UnionAll
+  , keyword "intersect" $> Sig.Intersect
+  , keyword "intersect" *> keyword "all" $> Sig.IntersectAll
```

Comment:
Just realized this is wrong too, it will only ever parse "intersect", and never "intersect all".